applications/serial_lte_modem/src/gnss/slm_at_gnss.c

```diff
@@ -0,0 +1,681 @@
+		return;
+	}
+	k_work_submit(&agps_req);
+#endif /* CONFIG_SLM_AGPS */
+
+#if defined(CONFIG_SLM_PGPS)
+	int err2;
+
+	err2 = nrf_cloud_pgps_request_all();
+	if (err2) {
+		LOG_ERR("Failed to request P-GPS data: %d", err2);
+	}
+	err2 = nrf_cloud_pgps_notify_prediction();
+	if (err2) {
+		LOG_ERR("Failed to request notification of prediction: %d", err2);
+	}
+#endif /* CONFIG_SLM_PGPS */
+}
+
+/* NOTE this event handler runs in interrupt context */
+static void gnss_event_handler(int event)
+{
+	switch (event) {
+	case NRF_MODEM_GNSS_EVT_PVT:
+		LOG_INF("NRF_MODEM_GNSS_EVT_PVT");
+		break;
+	case NRF_MODEM_GNSS_EVT_FIX: {
+		struct nrf_modem_gnss_pvt_data_frame pvt;
+		int err;
+
+		LOG_INF("NRF_MODEM_GNSS_EVT_FIX");
+		if (ttft_start != 0) {
+			LOG_INF("TTFF %ds", (int)k_uptime_delta(&ttft_start)/1000);
+			ttft_start = 0;
+		}
+		err = nrf_modem_gnss_read((void *)&pvt, sizeof(pvt), NRF_MODEM_GNSS_DATA_PVT);
+		if (err) {
+			LOG_ERR("Failed to read GNSS PVT data, error %d", err);
+			break;
+		}
+		sprintf(rsp_buf,
+			"\r\n#XGPS: %lf,%lf,%f,%f,%f,%f,\"%04u-%02u-%02u %02u:%02u:%02u\"\r\n",
+			pvt.latitude, pvt.longitude, pvt.altitude,
+			pvt.accuracy, pvt.speed, pvt.heading,
+			pvt.datetime.year, pvt.datetime.month, pvt.datetime.day,
+			pvt.datetime.hour, pvt.datetime.minute, pvt.datetime.seconds);
+		rsp_send(rsp_buf, strlen(rsp_buf));
+#if defined(CONFIG_SLM_PGPS)
+		struct tm gps_time = {
+			.tm_year = pvt.datetime.year - 1900,
+			.tm_mon  = pvt.datetime.month - 1,
+			.tm_mday = pvt.datetime.day,
+			.tm_hour = pvt.datetime.hour,
+			.tm_min  = pvt.datetime.minute,
+			.tm_sec  = pvt.datetime.seconds,
+		};
+		date_time_set(&gps_time);
+		nrf_cloud_pgps_set_location(pvt.latitude, pvt.longitude);
+#endif /* CONFIG_SLM_PGPS */
+	} break;
+	case NRF_MODEM_GNSS_EVT_NMEA:
+		LOG_INF("NRF_MODEM_GNSS_EVT_NMEA");
+		break;
+	case NRF_MODEM_GNSS_EVT_AGPS_REQ:
+		LOG_INF("NRF_MODEM_GNSS_EVT_AGPS_REQ");
+		on_gnss_evt_agps_req();
+		break;
+	case NRF_MODEM_GNSS_EVT_BLOCKED:
+		LOG_WRN("NRF_MODEM_GNSS_EVT_BLOCKED");
+		break;
+	case NRF_MODEM_GNSS_EVT_UNBLOCKED:
+		LOG_INF("NRF_MODEM_GNSS_EVT_UNBLOCKED");
+		break;
+	case NRF_MODEM_GNSS_EVT_PERIODIC_WAKEUP:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_PERIODIC_WAKEUP");
+		break;
+	case NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_SLEEP_AFTER_TIMEOUT");
+		break;
+	case NRF_MODEM_GNSS_EVT_SLEEP_AFTER_FIX:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_SLEEP_AFTER_FIX");
+		break;
+	case NRF_MODEM_GNSS_EVT_REF_ALT_EXPIRED:
+		LOG_DBG("NRF_MODEM_GNSS_EVT_REF_ALT_EXPIRED");
+		break;
+	default:
+		break;
+	}
+}
+
+#if defined(NRF_CLOUD_ACCESS)
+static void on_cloud_evt_ready(void)
+{
+	int err;
+	struct cloud_msg msg = {
+		.qos = CLOUD_QOS_AT_MOST_ONCE,
+		.endpoint.type = CLOUD_EP_STATE,
+		.buf = SERVICE_INFO_GPS,
+		.len = strlen(SERVICE_INFO_GPS)
+	};
+
+	/* Update nRF Cloud with GPS service info signifying GPS capabilities. */
+	err = cloud_send(nrf_cloud, &msg);
```

Comment:
this only needs to be done once.  you could add a flag to prevent unnecessary data transmission.

Reply:
OK, so for A-GPS or P-GPS or CELL_POS, any type, the notification is persistent on cloud side right? Just wonder whether we really need to send this or not, as the purpose of this AT commands is to get the location instead of tracking on nRF Cloud

Reply:
this is used by the front end on nrfcloud.com to display the map.
so it could be removed here if you aren't interested in providing that functionality.
otherwise, it only needs to be sent once.  it is persistent in the device's shadow.

Reply:
I'll make a new AT command to do this, and leave decision to user whether the device should be tracked on nRF Cloud or not.